File: scripts/investigation/verify.mjs
import assert from "node:assert/strict";
import { readFileSync } from "node:fs";
import { createRequire } from "node:module";
const root = new URL("../../", import.meta.url);
const require = createRequire(new URL("site/package.json", root));
const ts = require("typescript");
// Only transpile the pure search module. No Next server or paid calls are needed.
const source = readFileSync(new URL("site/app/lib/investigation/search.ts", root), "utf8");
const js = ts.transpileModule(source, { compilerOptions: { module: ts.ModuleKind.ES2022, target: ts.ScriptTarget.ES2020 } }).outputText;
const { rankSignals, cosineRank } = await import(`data:text/javascript;base64,${Buffer.from(js).toString("base64")}`);
const data = JSON.parse(readFileSync(new URL("site/data/investigation/search.json", root), "utf8"));
const fixture = JSON.parse(readFileSync(process.argv[2], "utf8"));
for (const c of fixture.cases) assert.deepEqual(rankSignals(data, c.scope, c.selected, c.minimum, c.wobble), c.expected, JSON.stringify({ ...c, expected: undefined }));
assert.equal(rankSignals(data, "both", [], 0, false).length, 0);
const pair = ["self_justification_via_intent", "sim_vs_real_deliberation"];
const both = rankSignals(data, "both", pair, 4, false, true);
assert.equal(both.length, 8);
assert.ok(both.every((r) => pair.every((g) => r.groups.includes(g))));
const bytes = readFileSync(new URL("site/data/investigation/vectors.f32", root));
const matrix = new Float32Array(bytes.buffer, bytes.byteOffset, bytes.byteLength / 4);
const meta = JSON.parse(readFileSync(new URL("site/data/investigation/semantic.json", root), "utf8"));
for (const c of fixture.cosines) {
  const actual = cosineRank(matrix, c.vector, meta.dimensions, meta.chunks.length, 50);
  assert.deepEqual(actual.map((r) => r.index), c.expected.map((r) => r.index));
  actual.forEach((r, i) => assert.ok(Math.abs(r.score - c.expected[i].score) < 1e-10));
}
assert.throws(() => cosineRank(matrix, [NaN], meta.dimensions, meta.chunks.length, 10));
console.log(`PASS: ${fixture.cases.length} original Python regex rankings, AND/empty filters, ${fixture.cosines.length} NumPy cosine rankings, invalid vector guard.`);

File: scripts/investigation/verify.py
#!/usr/bin/env python3
"""Offline parity fixtures from the original Python ranking implementation."""
import argparse
import hashlib
import itertools
import json
from pathlib import Path
import sys

import numpy as np

sys.dont_write_bytecode = True
ROOT = Path(__file__).resolve().parents[2]
sys.path.insert(0, str(Path(__file__).parent / "source"))
import misalignment_ui as ui

parser = argparse.ArgumentParser()
parser.add_argument("--output", type=Path, required=True)
args = parser.parse_args()
data = ROOT / "site/data/investigation"
manifest = json.loads((data / "manifest.json").read_text())
for name, expected in manifest["outputs"].items():
    assert hashlib.sha256((data / name).read_bytes()).hexdigest() == expected, name
raw = ROOT / "site/public/investigation/source/transcript.jsonl"
assert hashlib.sha256(raw.read_bytes()).hexdigest() == manifest["source_sha256"]
ui.MESSAGES, _ = ui.load_messages(raw)
copied = json.loads((data / "messages.json").read_text())
assert len(copied) == len(ui.MESSAGES) == 2064
for original, exported in zip(ui.MESSAGES, copied):
    assert all(exported[k] == v for k, v in original.items()), original["index"]
search = json.loads((data / "search.json").read_text())
ui.GROUPS = ui.compile_groups(search["definitions"])
ui.BONUSES = search["definitions"]["cooccurrence_bonus"]
ui.SUMMARIES = search["summaries"]
ui.WOBBLE = {int(k): v for k, v in search["wobble"].items()}
cases = []
subsets = [list(ui.GROUPS), ["self_justification_via_intent", "sim_vs_real_deliberation"], ["flag_then_rationalize"], []]
for scope, minimum, selected, wobble in itertools.product(["both", "thinking", "visible"], [0, 4, 8], subsets, [False, True]):
    rows = ui.ranked_payload(scope, minimum, set(selected), 0.25 if wobble else None)
    cases.append({"scope": scope, "minimum": minimum, "selected": selected, "wobble": wobble, "expected": [{"index": r["index"], "score": r["score"], "groups": r["matched_groups"], "excerpt": r["best_excerpt"], "summary": r["summary"], "wobble": r["wobble"]} for r in rows]})
meta = json.loads((data / "semantic.json").read_text())
matrix = np.fromfile(data / "vectors.f32", dtype="<f4").reshape(-1, meta["dimensions"])
cosines = []
for row in [0, 129, 139, 2000, len(matrix) - 1]:
    query = matrix[row].astype(np.float64)
    scores = matrix.astype(np.float64) @ (query / np.linalg.norm(query))
    order = np.argsort(-scores, kind="stable")[:50]
    cosines.append({"vector": query.tolist(), "expected": [{"index": int(i), "score": float(scores[i])} for i in order]})
args.output.write_text(json.dumps({"cases": cases, "cosines": cosines}))
print(f"Archive hashes and all 2064 messages verified; {len(cases)} regex cases and {len(cosines)} cosine fixtures written.")
